### DISCUSSION TITLE
跨平台

### DISCUSSION THREAD

**Discussion by FlysoftBeta** [LorisYounger/VPet#196] [Show and tell]
如题，WPF 不好跨平台，使用 wine 一类模拟软件有兼容性问题

我现在正尝试用 compose for desktop 实现，这样就可以实现完全跨平台了

开源地址：https://github.com/FlysoftBeta/VPet-Compose

目前实现的功能：

- [x] 待机动画
- [x] 根据音量播放动作
- [ ] 宠物自动移动
- [x] 仅头部可拖动
- [x] 头部拖动动画
- [x] 头部拖动动画拖动点

**Comment by freeze-dolphin** [comment]
看到大家都在尝试自制跨平台版本...于是我放下了手中的gtk-rs（开摆

**Comment by FlysoftBeta** [comment]
https://github.com/LorisYounger/VPet/assets/49718840/a3a1d789-99ff-4c02-8c0d-3799bed4dd69

可惜占用比wpf大很多

![图片](https://github.com/LorisYounger/VPet/assets/49718840/ab8ecc53-d41d-47d3-bca6-1433785a8da4)
![图片](https://github.com/LorisYounger/VPet/assets/49718840/d4ac8f0f-98ca-4894-a10a-282e408910de)

**Comment by freeze-dolphin** [comment]
其实我觉得桌宠这种东西还是最好用 native 一点的方式来实现比较好
谁都不想用一个占用大却只能用来看看的玩具
不可能为了桌宠放弃更重要的事情

**Comment by FlysoftBeta** [comment]
> 其实我觉得桌宠这种东西还是最好用 native 一点的方式来实现比较好 谁都不想用一个占用大却只能用来看看的玩具 不可能为了桌宠放弃更重要的事情

毕竟要跨平台，运行效率、资源占用方面的牺牲是难以避免的🤷‍♂️（刚才优化了一下，现在只需要70多m的内存了

**Comment by freeze-dolphin** [comment]
> > 其实我觉得桌宠这种东西还是最好用 native 一点的方式来实现比较好 谁都不想用一个占用大却只能用来看看的玩具 不可能为了桌宠放弃更重要的事情
> 
> 毕竟要跨平台，性能方面的牺牲是难以避免的🤷‍♂️

那看是什么层面上的跨平台...像jvm这种compile once run everywhere的确实

但是也可以做到分发不同平台的可执行文件，用 C 之类的

其实 .NET 也可以做到，只是 author 用了 WPF...这就没办法了

我记得 .NET 有跨平台的图形界面框架

---

我还是继续倒腾一下 gtk-rs 叭....

**Comment by FlysoftBeta** [comment]
> > > 其实我觉得桌宠这种东西还是最好用 native 一点的方式来实现比较好 谁都不想用一个占用大却只能用来看看的玩具 不可能为了桌宠放弃更重要的事情
> > 
> > 
> > 毕竟要跨平台，性能方面的牺牲是难以避免的🤷‍♂️
> 
> 那看是什么什么层面上的跨平台...像jvm这种compile once run everywhere的确实
> 
> 但是也可以做到分发不同平台的可执行文件，用 C 之类的
> 
> 其实 .NET 也可以做到，只是 author 用了 WPF...这就没办法了
> 
> 我记得 .NET 有跨平台的图形界面框架
> 
> 我还是继续倒腾一下 gtk-rs 叭....

.net 的好像都不是很成熟，不太清楚（很久没碰过c#这些语言了）

也许h5+tauri这种方案更好一点？虽然内存占用有点大（但还不算离谱），但是体积很小，有系统webview就能用

**Reply by NOBB2333** [reply]
.NET8  不是有MAUI 么，全平台还能手机上跑，除了打包时候打一套UI库进去，存储大，但是内存好像还可以的，有待研究，哈哈哈

**Comment by freeze-dolphin** [comment]
> .net 的好像都不是很成熟，不太清楚（很久没碰过c#这些语言了）

我在想，不会到时候冒出来各种 VPet 的实现吧

现在已经有一个 Swift 的了

**Comment by freeze-dolphin** [comment]
> 也许h5+tauri这种方案更好一点？虽然内存占用有点大（但还不算离谱），但是体积很小，有系统webview就能用

我依然保持原有的观点...

虽然我对偏底层语言的认知基本为零，平时也就是写写 kotlin 之类 jvm 平台的东西

但是我觉得这也算一个很好的学习机会吧...

**Comment by FlysoftBeta** [comment]
> > .net 的好像都不是很成熟，不太清楚（很久没碰过c#这些语言了）
> 
> 我在想，不会到时候冒出来各种 VPet 的实现吧
> 
> 现在已经有一个 Swift 的了

~~不得不说作者的代码真的挺乱（划掉~~，写的时候解析资源文件都很麻烦（~~我只能用暴力穷举目录名称匹配了~~

> 但是我觉得这也算一个很好的学习机会吧...

确实，但是感觉静态语言做插件对接可能会有些麻烦（？

**Comment by freeze-dolphin** [comment]
我倒还没有去仔细研究作者的实现（我还在复习我扔掉了快一年的rust！）

但是我大致看了一下他的结构，貌似我们只需要编写一个 VPet 的插件加载器就行了（？

因为本体其实就是一个 `0000-core` 插件

**Comment by freeze-dolphin** [comment]
> 但是感觉静态语言做插件对接可能会有些麻烦（？

参考 C 和 Lua 的关系...

**Comment by FlysoftBeta** [comment]
> 因为本体其实就是一个 `0000-core` 插件

我主要想兼容创意工坊里部分包含csharp代码的插件（多语言对接对我来说简直是噩梦（悲

**Comment by freeze-dolphin** [comment]
> 兼容创意工坊里部分包含代码的插件

这倒是个问题...

那其实应该直接用 c# 来写

java 和 c# 倒是不差很多（语法）

我倒是没有这个打算，感觉原生的 VPet 已经够看了

可能像不包含代码的插件可以支持一下

或者自己写一套插件系统

**Comment by FlysoftBeta** [comment]
> 可能像不包含代码的插件可以支持一下
> 
> 或者自己写一套插件系统

这两个感觉可以都做

**Comment by LorisYounger** [comment]
我个人比较推荐的是调用我写的Core类库然后改用C#跨平台UI库
这样我写的内容可以同时同步到不同平台,保证统一性同时也支持创意工坊/mod等各种东西

**Comment by fanxitong** [comment]
> > > > 其实我觉得桌宠这种东西还是最好用 native 一点的方式来实现比较好 谁都不想用一个占用大却只能用来看看的玩具 不可能为了桌宠放弃更重要的事情
> > > 
> > > 
> > > 毕竟要跨平台，性能方面的牺牲是难以避免的🤷‍♂️
> > 
> > 
> > 那看是什么什么层面上的跨平台... 像jvm这种compile once run everywhere的确实
> > 但是也可以做到分发不同平台的可执行文件，用 C 之类的
> > 其实 .NET 也可以做到，只是 author 用了 WPF... 这就没办法了
> > 我记得 .NET 有跨平台的图形界面框架
> > 我还是继续倒腾一下 gtk-rs 叭....
> 
> .net 的好像都不是很成熟，不太清楚（很久没碰过c#这些语言了）
> 
> 也许h5+tauri这种方案更好一点？ 虽然内存占用有点大（但还不算离谱），但是体积很小，有系统webview就能用

.net上的跨平台目前还是[Avalonia](https://github.com/AvaloniaUI/Avalonia)好一点和WPF写法也相近。
如果作者能把Core库目标改为_.net standard 2.0_并将UI控件部分移到其它库中,只在Core库里保留操作逻辑，那么后边改动要好很多.

**Comment by LorisYounger** [comment]
> .net上的跨平台目前还是[Avalonia](https://github.com/AvaloniaUI/Avalonia)好一点和WPF写法也相近。 如果作者能把Core库目标改为_.net standard 2.0_并将UI控件部分移到其它库中,只在Core库里保留操作逻辑，那么后边改动要好很多.

下次一定, 等更新到1.1版本的时候会大改下调用方式, 把显示独立出来 具体见 #11

**Comment by FlysoftBeta** [comment]
开源地址：https://github.com/FlysoftBeta/VPet-Compose

**Comment by justaLoli** [comment]
原生开发的话，我用swift在mac上内存占用能保持在30m以下

![CleanShot 2023-09-11 at 16 23 14](https://github.com/LorisYounger/VPet/assets/92238102/98bb8487-2c69-44e8-9b8c-706bfb79b2b2)

**Reply by freeze-dolphin** [reply]
你这个还原度很高啊qwq

**Comment by FlysoftBeta** [comment]
很好奇 vup.lps 里面touchhead是怎么判定的，总体积是多少
![图片](https://github.com/LorisYounger/VPet/assets/49718840/081ec9a2-172b-465c-be47-88f5fa04d48b)

**Reply by FlysoftBeta** [reply]
经过测试，应该是500x500

**Comment by FlysoftBeta** [comment]
支持头部拖拽+拖拽动画
![图片](https://github.com/LorisYounger/VPet/assets/49718840/867d506b-2129-43bd-afb4-0626e3fb7223)

**Comment by ReallocAll** [comment]
在下图配置的情况下使用Steam提供的Proton运行VPet可获得良好的体验，各类窗口均可在正常打开，除了背景是纯黑，占用有点高以外（CPU：10%+，MEM：~800MB），基本没有问题
![image](https://github.com/LorisYounger/VPet/assets/63277109/3100a8c9-b8b0-40b6-a30d-1422b8b9f0d1)
![image](https://github.com/LorisYounger/VPet/assets/63277109/a62e2abd-310e-4c7f-b19b-2a508034749a)
![截图 2023-10-01 00-07-46](https://github.com/LorisYounger/VPet/assets/63277109/2c2ece74-fb2e-40e0-b688-338ce720c8bf)
